### PR TITLE
Johnsonshi/refactor azure report ready codepath

### PR DESCRIPTION
**Pull request by johnsonshi**
Refactor Azure report ready and Goal State code path.

### CODE REVIEW

**Comment by johnsonshi** [comment]
Hi @OddBloke, as discussed the closed PR #422, I have removed the diagnostic/error logging changes to the PR. The code changes now solely focus on refactoring the Azure report ready codepath and Goal State related code. Thanks for the suggestions!

**Comment by johnsonshi** [comment]
Adding @anhvoms as well for visibility.

File: cloudinit/sources/helpers/azure.py [hunk 1]
```diff
@@ -213,24 +213,70 @@ def get(self, url, secure=False):
...
+        self._instance_id = self._text_from_xpath(
+            './Container/RoleInstanceList/RoleInstance/InstanceId')
+        self._incarnation = self._text_from_xpath('./Incarnation')
+
+        if any(x is None for x in (self.container_id,
+                                   self.instance_id, self.incarnation)):
+            msg = ('Missing container id/instance id/incarnation in '
+                   'GoalState XML.')
+            LOG.warning(msg)
+            report_diagnostic_event(msg)
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: I will fix these repetitions in a follow-up PR immediately after this as discussed :)

File: cloudinit/sources/helpers/azure.py [hunk 2]
```diff
@@ -213,24 +213,70 @@ def get(self, url, secure=False):
...
+        @return: GoalState object representing the GoalState XML string.
+        """
+        self.azure_endpoint_client = azure_endpoint_client
+
+        try:
+            self.root = ElementTree.fromstring(unparsed_xml)
+        except ElementTree.ParseError as e:
+            msg = 'Failed to parse GoalState XML: %s' % e
+            LOG.warning(msg)
+            report_diagnostic_event(msg)
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: I will fix these repetitions in a follow-up PR immediately after this as discussed :)

File: cloudinit/sources/helpers/azure.py [hunk 3]
```diff
@@ -240,26 +286,18 @@ def _text_from_xpath(self, xpath):
 
     @property
     def container_id(self):
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: Great idea! Incorporated your suggestion and set the un-underscored attributes directly in `__init__`.

File: cloudinit/sources/helpers/azure.py [hunk 4]
```diff
@@ -536,42 +669,131 @@ def find_endpoint(fallback_lease_file=None, dhcp245=None):
...
+        @param azure_endpoint_client: instance of AzureEndpointHttpClient
+        @return: GoalState object representing the GoalState XML
+        """
+        unparsed_goal_state_xml = self._get_goal_state_from_azure(
+            azure_endpoint_client)
+        return self._parse_goal_state(
+            unparsed_goal_state_xml, azure_endpoint_client)
+
+    @azure_ds_telemetry_reporter
+    def _get_goal_state_from_azure(self, azure_endpoint_client):
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: Thanks for the suggestion. When I looked at the method names again, they indeed are confusing. I've renamed them to `_fetch_goal_state_from_azure` (which returns a `GoalState` object) and `_get_raw_goal_state_xml_from_azure` (which returns the raw goal state xml string). Let me know if these names look good.

File: cloudinit/sources/helpers/azure.py [hunk 5]
```diff
@@ -536,42 +669,131 @@ def find_endpoint(fallback_lease_file=None, dhcp245=None):
...
+        """Fetches the GoalState XML from the Azure endpoint and returns
+        the XML as a string.
+
+        @param azure_endpoint_client: instance of AzureEndpointHttpClient
+        @return: GoalState XML string
+        """
+
         LOG.info('Registering with Azure...')
-        attempts = 0
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py:543]: Yes retries are done using `readurl`'s args: `url_helper.readurl(url, headers=headers, timeout=5, retries=10, sec_between=5)`

File: cloudinit/sources/helpers/azure.py [hunk 6]
```diff
@@ -536,42 +669,131 @@ def find_endpoint(fallback_lease_file=None, dhcp245=None):
...
+            self.openssl_manager.certificate)
+        goal_state = self._fetch_goal_state_from_azure(azure_endpoint_client)
+        ssh_keys = self._get_user_pubkeys(goal_state, pubkey_info)
+        health_reporter = GoalStateHealthReporter(
+            goal_state, azure_endpoint_client, self.endpoint)
+        health_reporter.send_ready_signal()
+        return {'public-keys': ssh_keys}
+
+    @azure_ds_telemetry_reporter
+    def _fetch_goal_state_from_azure(self, azure_endpoint_client):
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: Great suggestion. Done! :)

File: cloudinit/sources/helpers/azure.py [hunk 7]
```diff
@@ -370,25 +408,105 @@ def parse_certificates(self, certificates_xml):
...
 
 
-class WALinuxAgentShim(object):
+class GoalStateHealthReporter(object):
+
+    HEALTH_REPORT_XML_TEMPLATE = textwrap.dedent('''\
+        <?xml version="1.0" encoding="utf-8"?>
+        <Health xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+          <GoalStateIncarnation>{incarnation}</GoalStateIncarnation>
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py:402]: Thanks for pointing out your concerns. The `incarnation`, `container_id`, and `instance_id` portions of this XML each are short strings obtained from the Azure endpoint. These three are obtained from the `GoalState` xml from the platform, and these values are fully controlled by the Azure platform and the GoalState channel is trusted.

The `health_status` and `health_detail_subsection` are fully controlled by cloud-init code and no untrusted actor would be able to mutate these. As of now, `health_status` is always equal to `Ready` (reporting to the Azure platform that the VM is in a healthy state), while `health_detail_subsection` is always empty (for now).

In a PR after the subsequent PR for Azure reporting/logging, I will be introducing logic for cloud-init to report non-Ready status (non-healthy status) to the Azure platform. The future report non-healthy logic is to allow early termination of the VM deployment if cloud-init encounters an irrecoverable error during VM deployment on Azure (such as platform/host issues where the VM was deployed in). In such non-healthy cases, `health_status` and `health_detail_subsection` will contain different values, but they will still be fully-controlled within cloud-init code as constants (without external control).

Thanks for your concern! :)

File: cloudinit/sources/helpers/azure.py [hunk 8]
```diff
@@ -370,25 +408,105 @@ def parse_certificates(self, certificates_xml):
...
+            container_id=container_id,
+            instance_id=instance_id,
+            health_status=status,
+            health_detail_subsection=health_detail)
+
+        return health_report
+
+    @azure_ds_telemetry_reporter
+    def _post_health_report(self, document):
+        """Host will collect kvps when cloud-init reports to fabric.
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: Intended to be a comment for the `time.sleep(0)` line. I've since updated the comments and added extra explanation for the code. Let me know if it looks good :)

File: cloudinit/sources/helpers/azure.py [hunk 9]
```diff
@@ -370,25 +408,105 @@ def parse_certificates(self, certificates_xml):
...
+        Some kvps might still be in the queue. We yield the scheduler
+        to make sure we process all kvps up till this point.
+        """
+        time.sleep(0)
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py:498]: I've since updated the comments and added extra explanation for the code. Let me know if it looks good :)

File: cloudinit/sources/helpers/azure.py [hunk 10]
```diff
@@ -536,42 +669,131 @@ def find_endpoint(fallback_lease_file=None, dhcp245=None):
...
+        LOG.debug(msg)
+        report_diagnostic_event(msg)
+        return goal_state
+
+    @azure_ds_telemetry_reporter
+    def _get_user_pubkeys(self, goal_state, pubkey_info):
+        """Gets and filters the VM user's authorized pubkeys.
+
+        cloud-init expects a straightforward array of keys to be dropped
+        into the user's authorized_keys file. Azure control plane exposes
```
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py]: The "user" here is the "admin user" that is specified when a VM is deployed on Azure. See https://docs.microsoft.com/en-us/cli/azure/vm?view=azure-cli-latest#az-vm-create and https://docs.microsoft.com/en-us/cli/azure/vm#az-vm-create.

I have since updated the docs to reflect this.

**Comment by johnsonshi** [comment]
Hi @OddBloke, thanks for the very detailed and thorough review! I have followed the mock test naming conventions you mentioned earlier, your type annotation PR comments, refactoring `AzureEndpointHttpClient` to be in `self` (instead of getting passed around), and a bunch of other PR comments. Let me know if you need anything else.

**Comment by OddBloke** [comment]
Hey @johnsonshi, thanks for the update! I was off at the end of last week, so I'm playing catch-up; I'm expecting to get to a re-review of this tomorrow.

File: tests/unittests/test_datasource/test_azure_helper.py [hunk 11]
```diff
@@ -365,6 +459,113 @@ def test_parse_certificates(self, mock_decrypt_certs):
...
+            azure_helper.GoalState(mock.MagicMock(), mock.MagicMock()),
+            azure_helper.AzureEndpointHttpClient(mock.MagicMock()),
+            self.test_endpoint)
+        generated_health_document = reporter.build_report(
+            incarnation=self.default_parameters['incarnation'],
+            container_id=self.default_parameters['container_id'],
+            instance_id=self.default_parameters['instance_id'],
+            status=self.provisioning_success_status)
+        self.assertEqual(health_document, generated_health_document)
+        self.assertIn(str(self.default_parameters['incarnation']),
```
Comment (OddBloke) [inline tests/unittests/test_datasource/test_azure_helper.py:537]: This is just testing that "1" appears in the generated document, somewhere, I think? The first line of the template includes "1" so this test will always pass, regardless of whether or not the incarnation value is used.

I suggest either using a large number that is unlikely to appear in the template in future or, probably better, asserting that the full XML element is present (i.e. test for `<GoalStateIncarnation>1</GoalStateIncarnation>`).

(The `container_id` and `instance_id` assertions don't have the same problem as they are using unique strings, but it would still be good to assert that the XML element is correct.)

File: tests/unittests/test_datasource/test_azure_helper.py [hunk 12]
```diff
@@ -365,6 +459,113 @@ def test_parse_certificates(self, mock_decrypt_certs):
...
+
+    def _get_report_ready_health_document(self):
+        return self._get_formatted_health_report_xml_string(
+            incarnation=self.default_parameters['incarnation'],
+            container_id=self.default_parameters['container_id'],
+            instance_id=self.default_parameters['instance_id'],
+            health_status=self.provisioning_success_status,
+            health_detail_subsection='')
+
+    def test_send_ready_signal_sends_post_request(self):
```
Comment (OddBloke) [inline tests/unittests/test_datasource/test_azure_helper.py:510]: I feel like there's some overlap in these tests which could be cleaned up to make them a little clearer. This test implies it tests that "`send_ready_signal` sends a POST request", but it's also testing that `build_report` is called and that `build_report` behaves as expected; if not, the final assertion would fail here because `data=health_document` would not match. This encompasses the following two tests entirely, rendering them unnecessary.

`test_send_ready_signal_health_document` is already testing `build_report`, so we don't also need to test it here. I propose modifying this test to include the mocking of `build_report` from `test_send_ready_signal_calls_build_report`, and instead of `mock.call(..., data=health_document, ...)` instead test for `mock.call(..., data=m_build_report.return_value, ...)`. (I suggest keeping `test_send_ready_signal_calls_build_report` separate; validating the parameters can be performed separately, and the less each test is asserting, the easier it is to grok them.)

File: tests/unittests/test_datasource/test_azure_helper.py [hunk 13]
```diff
@@ -365,6 +459,113 @@ def test_parse_certificates(self, mock_decrypt_certs):
...
+        reporter.send_ready_signal()
+        self.assertEqual(1, self.post.call_count)
+        self.assertEqual(
+            mock.call(
+                self.test_url,
+                data=health_document,
+                extra_headers=self.test_default_headers),
+            self.post.call_args)
+
+    def test_send_ready_signal_health_document(self):
```
Comment (OddBloke) [inline tests/unittests/test_datasource/test_azure_helper.py:525]: This isn't testing `send_ready_signal`, it's testing `build_report`.

**Comment by johnsonshi** [comment]
Looking at the Hyper-V KVP Reporter's threaded publisher subroutine ([here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L140-L144) and [here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L291)), we see that it calls [`_append_kvp_item`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L231-L237) in order to write `encoded_data` (encoded kvp diagnostic messages) to the kvp pool file.

We see that `_append_kvp_item` does hold an exclusive lock and flushes output to the file, so we can be sure that diagnostic messages are indeed flushed to the Hyper-V kvp file.

We also see that [`HyperVKvpReportingHandler.flush()`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L325-L327) does block the main thread until the publisher thread completely empties queued kvp diagnostic messages.

You are right that `time.sleep(0)` is not needed since (1) the threaded publishing subroutine flushes output when it writes and (2) the main thread blocks until threaded publisher empties and writes all queued events. Thus we can be sure that all queued events are written up to this point.

In fact, it seems like `HyperVKvpReportingHandler.flush()` is a reliable method to ensure that queued events are indeed empties and flushed. The previous `time.sleep(0)` did not provide any guarantees of flushing all events. In fact it was just relying on yielding to the task scheduler and random luck. :)

**Comment by johnsonshi** [comment]
Thank you so much @OddBloke. I can sense that this is getting really close indeed. Thanks for your extensive reviews each pass :)

**Comment by anhvoms** [comment]
> Looking at the Hyper-V KVP Reporter's threaded publisher subroutine ([here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L140-L144) and [here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L291)), we see that it calls [`_append_kvp_item`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L231-L237) in order to write `encoded_data` (encoded kvp diagnostic messages) to the kvp pool file.
> 
> We see that `_append_kvp_item` does hold an exclusive lock and flushes output to the file, so we can be sure that diagnostic messages are indeed flushed to the Hyper-V kvp file.
> 
> We also see that [`HyperVKvpReportingHandler.flush()`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L325-L327) does block the main thread until the publisher thread completely empties queued kvp diagnostic messages.
> 
> You are right that `time.sleep(0)` is not needed since (1) the threaded publishing subroutine flushes output when it writes and (2) the main thread blocks until threaded publisher empties and writes all queued events. Thus we can be sure that all queued events are written up to this point.
> 
> In fact, it seems like `HyperVKvpReportingHandler.flush()` is a reliable method to ensure that queued events are indeed empties and flushed. The previous `time.sleep(0)` did not provide any guarantees of flushing all events. In fact it was just relying on yielding to the task scheduler and random luck. :)



> Looking at the Hyper-V KVP Reporter's threaded publisher subroutine ([here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L140-L144) and [here](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L291)), we see that it calls [`_append_kvp_item`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L231-L237) in order to write `encoded_data` (encoded kvp diagnostic messages) to the kvp pool file.
> 
> We see that `_append_kvp_item` does hold an exclusive lock and flushes output to the file, so we can be sure that diagnostic messages are indeed flushed to the Hyper-V kvp file.
> 
> We also see that [`HyperVKvpReportingHandler.flush()`](https://github.com/canonical/cloud-init/blob/ebc145be5e15c64a31ba496625727a7308a57baf/cloudinit/reporting/handlers.py#L325-L327) does block the main thread until the publisher thread completely empties queued kvp diagnostic messages.
> 
> You are right that `time.sleep(0)` is not needed since (1) the threaded publishing subroutine flushes output when it writes and (2) the main thread blocks until threaded publisher empties and writes all queued events. Thus we can be sure that all queued events are written up to this point.
> 
> In fact, it seems like `HyperVKvpReportingHandler.flush()` is a reliable method to ensure that queued events are indeed empties and flushed. The previous `time.sleep(0)` did not provide any guarantees of flushing all events. In fact it was just relying on yielding to the task scheduler and random luck. :)

HyperVKvpReportingHandler.flush() is reliable since it calls a queue join, but you risk blocking the report ready if the KVPs could not be flushed due to some reasons. I ran into this once when I had some malformed KVP due to utf-8 encoding. Regarldess, my opinion is that the telemetry channel should not fail the VM reporting ready. There could be bugs in the telemetry channel that we should figure out to make sure it would never cause issues, but that should not be the scope of this PR.

What we observed was that at the time of reporting ready, sometimes there were KVPs that weren't processed yet in the queue. Time.sleep(0) was a low cost best-effort to yield the scheduler so that some of these events could be processed. Is it random luck? I'm not quite sure. When i looked into this the last time, running strace showed that time.sleep(0) did trigger a context switch and schedule yielding. 
https://stackoverflow.com/questions/7273474/behavior-of-pythons-time-sleep0-under-linux-does-it-cause-a-context-switch

**Comment by johnsonshi** [comment]
After discussing with @anhvoms, I've decided to revert `time.sleep(0)` back as it was a proven and reliable method for KVPs to be written to disk before the VM reports ready. Agree with @anhvoms and his points. This PR should be scoped to refactoring only.

**Comment by johnsonshi** [comment]
@OddBloke I believe this is ready for merging now :)

File: tests/unittests/test_datasource/test_azure_helper.py [hunk 14]
```diff
@@ -365,6 +459,133 @@ def test_parse_certificates(self, mock_decrypt_certs):
...
+            container_id=self.default_parameters['container_id'],
+            instance_id=self.default_parameters['instance_id'],
+            status=self.provisioning_success_status)
+        self.assertEqual(health_document, generated_health_document)
+        self.assertIn(
+            ''.join([
+                '<GoalStateIncarnation>',
+                str(self.default_parameters['incarnation']),
+                '</GoalStateIncarnation>']),
+            generated_health_document)
```
Comment (OddBloke) [inline tests/unittests/test_datasource/test_azure_helper.py:545]: We can rewrite these assertions to express a little more clearly what it is they are testing:

```suggestion
        self.assertIn(
                '<GoalStateIncarnation>{}</GoalStateIncarnation>'.format(
                    str(self.default_parameters['incarnation'])),
                generated_health_document)
```

**Comment by johnsonshi** [comment]
@OddBloke As per @anhvoms 's comment above, there were issues with Hyper-V's flush KVP method due to encoding errors, causing the telemetry channel to be blocked and preventing cloud-init from ever flushing the KVPs. We'd prefer to scope this PR to only be refactoring, and have the next PR deal with telemetry/reporting/logging/flushing KVPs as previously discussed :)

**Comment by johnsonshi** [comment]
Also @OddBloke should the type hints in this PR be removed? Type hints were introduced in Python 3.5, and cloud-init still supports Python 3.4 I believe. Additionally, older distros such as RHEL/CentOS (which only ships with Python 2.7) might not be able to easily backport changes that are Python 3 specific. What's your take on this?

**Comment by OddBloke** [comment]
> @OddBloke As per @anhvoms 's comment above, there were issues with Hyper-V's flush KVP method due to encoding errors, causing the telemetry channel to be blocked and preventing cloud-init from ever flushing the KVPs. We'd prefer to scope this PR to only be refactoring, and have the next PR deal with telemetry/reporting/logging/flushing KVPs as previously discussed :)

I'm perfectly happy with that approach. If we are going to deal with such things in a follow-up PR, then we should drop the `time.sleep(0)` call: I don't want us to land only part of the full change which you intend to make.

File: cloudinit/sources/helpers/azure.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (johnsonshi) [reply cloudinit/sources/helpers/azure.py:495]: We'd like to scope this PR to only be refactoring and not include telemetry changes.

**Comment by johnsonshi** [comment]
Just to clarify @OddBloke, the `time.sleep(0)` call was previously there (currently in master). Then there was a proposal to change the `time.sleep(0)` to use flush KVP instead. However, after consulting with @anhvoms, we discovered that he did try to use flush KVP but ran into issues, so I reverted back to using `time.sleep(0)`, which was the original implementation.

**Comment by OddBloke** [comment]
> Also @OddBloke should the type hints in this PR be removed?

TL;DR: no. :-)

> Type hints were introduced in Python 3.5, and cloud-init still supports Python 3.4 I believe.

Function annotations have been supported in every version of Python 3: https://www.python.org/dev/peps/pep-3107/

What was introducing in Python 3.5 was the `typing` module, which our hacking doc prohibits the use of (for the compatibility reason you identified); see https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#type-annotations for more details.

> Additionally, older distros such as RHEL/CentOS (which only ships with Python 2.7) might not be able to easily backport changes that are Python 3 specific. What's your take on this?

cloud-init upstream is Python 3-only, we have already introduced changes which break Python 2 compatibility (including function annotations), and our codebase is better with type annotations than without: I don't think we should drop them.

**Comment by OddBloke** [comment]
> Just to clarify @OddBloke, the `time.sleep(0)` call was previously there (currently in master). Then there was a proposal to change the `time.sleep(0)` to use flush KVP instead. However, after consulting with @anhvoms, we discovered that he did try to use flush KVP but ran into issues, so I reverted back to using `time.sleep(0)`, which was the original implementation.

Aha, I see that now: thanks for pushing back! Given that, we should certainly leave it in.
